src/main/java/de/dhbw/music/ShapeProcessor.java

```diff
@@ -0,0 +1,119 @@
+        // TODO > treat control markers
+    }
+
+    public boolean[][] getSoundMatrix(){
+        return soundMatrix;
+    }
+
+    private void detectPlayfield(){
+        playfieldInfo[4] = 1;
+        Shape[] squares = shapes.stream().filter(shape -> shape.getForm() == ShapeForm.SQUARE).toArray(Shape[]::new);
+        if(squares.length < 4){
+            playfieldInfo[4] = 0;
+            System.err.println("Not enough squares for playfield");
+        }
+        else {
+            Shape[] corners = new Shape[]{squares[0], squares[1], squares[2], squares[3]};
+            for (Shape square : squares) {
+                // top left
+                if(Math.sqrt(square.pos[1] * square.pos[1] + square.pos[0] * square.pos[0]) < Math.sqrt(corners[0].pos[0] * corners[0].pos[0] + corners[0].pos[1] * corners[0].pos[1])){
+                    corners[0] = square;
+                }
+                // bottom right
+                if(Math.sqrt(square.pos[1] * square.pos[1] + square.pos[0] * square.pos[0]) > Math.sqrt(corners[2].pos[0] * corners[2].pos[0] + corners[2].pos[1] * corners[2].pos[1])){
+                    corners[2] = square;
+                }
+                // top right
+                if(Math.sqrt( square.pos[0] * square.pos[0] + (frameHeight - square.pos[1]) * (frameHeight - square.pos[1])) > Math.sqrt(corners[1].pos[0] * corners[1].pos[0] + (frameHeight - corners[1].pos[1]) * (frameHeight - corners[1].pos[1]))){
+                    corners[1] = square;
+                }
+                // bottom left
+                if(Math.sqrt((frameWidth - square.pos[0]) * (frameWidth - square.pos[0]) + square.pos[1] * square.pos[1]) > Math.sqrt((frameWidth - corners[3].pos[0]) * (frameWidth - corners[3].pos[0]) + corners[3].pos[1] * corners[3].pos[1])){
+                    corners[3] = square;
+                }
+            }
+            for(Shape c : corners) c.setType(ShapeType.FIELD_MARKER);
+            for(int i = 0; i < 4; i++){
+                playFieldBoundaries[i].put(0,0, corners[i].pos[0] - corners[(i+1) % 4].pos[0]);
+                playFieldBoundaries[i].put(1,0, corners[i].pos[1] - corners[(i+1) % 4].pos[1]);
+            }
+            playfieldInfo[0] = (corners[0].pos[0] + corners[3].pos[0]) / 2;
+            playfieldInfo[1] = (corners[0].pos[1] + corners[1].pos[1]) / 2;
+            playfieldInfo[2] = (corners[1].pos[0] + corners[2].pos[0]) / 2;
+            playfieldInfo[3] = (corners[2].pos[1] + corners[3].pos[1]) / 2;
+            playfieldInfo[4] = checkRectangularity() ? 1 : 0;
+            for(Shape s : shapes){
+                if(s.pos[0] > playfieldInfo[0] && s.pos[0] < playfieldInfo[0] + playfieldInfo[2]
+                        && s.pos[1] > playfieldInfo[1] && s.pos[1] < playfieldInfo[1] + playfieldInfo[3]){
+                    s.setType(ShapeType.SOUND_MARKER);
+                }
+            }
+        }
+    }
+
+    /**
+     * This method must not be called when the playfield is not valid!!!
```

Comment:
Intuitiv würde ich das ja in der Methode noch mal validieren. Ich schätze das hast du bewusst nicht gemacht, um Performance zu gewinnen?

Reply:
sí